### PR TITLE
Update product.id_category_default after deleting categories

### LINKED ISSUE #30219: Default category should always be enforced on a product, even after category is deleted
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

comes from PR https://github.com/PrestaShop/PrestaShop/pull/28141#issuecomment-1303654813

When deleting category, it is possible to leave certain products without default category, which leads to warning in FO and possibly other issues.
Currently the DeleteCategoryHandler only update product categories when they have no categories left, but it doesn't update those that have no default category left. (the product.id_default_category ends up invalid because it points to category which doesn't exist anymore).

### Expected behavior

when deleting category, we should also update products that doesn't have the default category left (or the ones that the default category was the one that we just deleted), so that its not possible for product to end up without default category.

### Steps to reproduce

1. Create 2 categories (I called it catD, catE)
2. Go to BO product page (no matter v1 or v2). Select any product for edition.
3. Add catD and catE to the product (so there is no other categories for that product), and mark catD as default one.
Screenshot from 2022-11-04 16-28-49

4. Go BO categories page and delete catD (I chose first option when deleting, but it shouldn't matter)
5. Go to BO product page (same product) and see it doesn't have the default category anymore.
Screenshot from 2022-11-04 16-29-26
6. Go to FO, find your product, open it and see error mentioning `Trying to access array offset on value of type bool in /var/www/html/prestashop/classes/Category.php on line 2018`
Screenshot from 2022-11-04 16-29-42

### PrestaShop version(s) where the bug happened

develop (but pretty sure its the same with 178 and 8.0 and maybe even some earlier versions)

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Introduced new method in AbstractDeleteCategoryHandler which finds all products that has id_category_default amongst the recently deleted categories. For each of these products it finds parent category (parent of the one that was deleted, or fallbacks to a PS_HOME_CATEGORY) and assigns it, so that product doesn't end up being without default category. This method is called after handling the products which have no categories assigned at all (refactored a bit, but that ones logic doesn't change). Also it is moved out of a loop in bulk action, so it doesn't need to fetch the categories over and over again. Added behat tests (and refactored old tests which were really bad)
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yes (see bellow)
| Fixed ticket?     | Fixes #30219
| Related PRs       | 
| How to test?      | See bellow
| Possible impacts? | 

**deprecated** `PrestaShop\PrestaShop\Adapter\Category\CommandHandler\AbstractDeleteCategoryHandler::handleProductsUpdate`, Use `updateProductCategories` instead

**how to test**
Note - this scenario only applies if after category deletion there are other categories still left assigned with the product, but product ends up having no default category. In other case - If after category deletion the product is left without categories, then the "deletion mode" behavior applies as usually.

1. Create new categories "catA and catB".
2. Go to any product and assign those 2 categories and make catA default one. Make sure product now has only those 2 categories assigned.
3. Go to BO categories page and delete catA (no matter which deletion mode).
4. Go back to product page (v1 or v2 should both act same) and see that instead of the deleted "catA" the product default category is the parent category of the "catA" now

### CODE REVIEW

File: src/Adapter/Category/CommandHandler/BulkDeleteCategoriesHandler.php [hunk 1]
```diff
@@ -62,7 +63,13 @@ public function handle(BulkDeleteCategoriesCommand $command)
                 throw new FailedToDeleteCategoryException(sprintf('Failed to delete category with id %s', var_export($categoryId->getValue(), true)));
             }
 
-            $this->handleProductsUpdate((int) $category->id_parent, $command->getDeleteMode());
+            $deletedCategoryIdsByParent[(int) $category->id_parent] = [$categoryId->getValue()];
```
Comment (jolelievre) [inline src/Adapter/Category/CommandHandler/BulkDeleteCategoriesHandler.php]: Careful here because you don't if the parent you are storing here may be one of the deleted categories (or one of their children) You need to make sure that all these parent IDs still match existing categories

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 2]
```diff
@@ -72,4 +76,140 @@ protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $m
...
+     *
+     * @return int|null
+     */
+    private function findCategoryParentId(int $categoryId, array $deletedCategoryIdsByParent): ?int
+    {
+        $parentCategoryId = null;
+        foreach ($deletedCategoryIdsByParent as $parentId => $deletedIds) {
+            // find parent id for deleted category
+            if (in_array($categoryId, $deletedIds, true)) {
+                $parentCategoryId = $parentId;
```
Comment (jolelievre) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
                return $parentId;
```

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 3]
```diff
@@ -72,4 +76,140 @@ protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $m
...
+                break;
+            }
+        }
+
+        return $parentCategoryId;
```
Comment (jolelievre) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
        return null;
```

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 4]
```diff
@@ -72,4 +76,140 @@ protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $m
...
+
+    /**
+     * @param int $categoryId
+     * @param array<int, int[]> $deletedCategoryIdsByParent
+     *
+     * @return int|null
+     */
+    private function findCategoryParentId(int $categoryId, array $deletedCategoryIdsByParent): ?int
+    {
+        $parentCategoryId = null;
```
Comment (jolelievre) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
```

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 5]
```diff
@@ -72,4 +76,140 @@ protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $m
...
+                continue;
+            }
+
+            if ($mode->shouldRemoveProducts()) {
+                $product->delete();
+
+                continue;
+            }
+
+            if ($mode->shouldDisableProducts()) {
```
Comment (jolelievre) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php:191]: So it means this mode was never handled correctly?

Comment (zuk3975) [reply src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php:191]: Hmm I actually don't understand why would we want to delete product if parent category id = 0 :thinking: 
The only legit case when category doesn't have parent is when its root category (isnt it?). But why would we want to delete such product?

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 6]
```diff
@@ -412,6 +414,20 @@ public function searchCombinations(
...
+        $result = $this->connection->createQueryBuilder()
+            ->addSelect('p.id_tax_rules_group')
+            ->from($this->dbPrefix . 'product', 'p')
+            ->where('id_product = :productId')
+            ->setParameter('productId', $productId->getValue())
+            ->execute()
+            ->fetchOne()
+        ;
+
+        return new TaxRulesGroupId((int) $result);
```
Comment (lartist) [inline src/Adapter/Product/Repository/ProductRepository.php:426]: $result is from `fetchOne`method.
-> It could be false.
-> When you cast false the result is 0.
`TaxRulesGroupId` only assert if the id is positif `0 > $value`
->Then you can have a value set to 0 although this Id doesn't exist.

Comment (zuk3975) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: Mhm I see. This change came from other PR after rebase. I only moved the method up now (cuz its public and should be above private methods). But Im afraid modifying it as It is out of scope.
@FabienPapet I see its your change, I wonder must product always have a taxRulesGroupId? It might be a mistake in a contract (it may instead be "findProductTaxRuleId" with a nullable return or else it should throw TaxRuleNotFoundException) , could you check it and fix in a dedicated PR?

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php (hunk 5 above)
Comment (jolelievre) [reply src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php:191]: 🤷‍♂️ not sure but in a way it makes sense, let's say you stop selling clothes you remove the Category clothes and all its associated products I guess.

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 7]
```diff
@@ -547,17 +563,53 @@ private function loadProduct(Product $product): Product
...
-            ->fetchOne()
-        ;
+        $results = Db::getInstance()->executeS('
+			SELECT p.`id_product`
+			FROM `' . _DB_PREFIX_ . 'product` p
+			' . Shop::addSqlAssociation('product', 'p') . '
+			WHERE NOT EXISTS (
+			    SELECT 1 FROM `' . _DB_PREFIX_ . 'category_product` cp WHERE cp.`id_product` = p.`id_product`
+			)
+		');
```
Comment (jolelievre) [inline src/Adapter/Product/Repository/ProductRepository.php]: Please replace this one with a query relying on `$this->connection` Let's avoid introducing `Db` in those recent services If the query builder is too complicated to handle you can always create a native SQL query.

Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductRepository.php]: Although I think you can use this recently merged trait by @FabienPapet https://github.com/FabienPapet/PrestaShop/blob/e85338ae4e79ce26bcda155d2dd57fd4824c53d8/src/Core/Repository/ShopConstraintTrait.php

Which basically allows to replace the `Shop::addSqlAssociation` in a query builder based on a `ShopConstraint`

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 8]
```diff
@@ -547,17 +563,53 @@ private function loadProduct(Product $product): Product
...
 
-        return new TaxRulesGroupId((int) $result);
+    /**
+     * @param int[] $defaultCategoryIds
+     *
+     * @return ProductId[]
+     */
+    public function findProductIdsByDefaultCategories(array $defaultCategoryIds): array
+    {
+        $results = Db::getInstance()->executeS('
```
Comment (jolelievre) [inline src/Adapter/Product/Repository/ProductRepository.php]: SAme suggestion, don't use Db anywhere in this file

File: src/Core/Domain/Category/Command/EditRootCategoryCommand.php [hunk 9]
```diff
@@ -31,6 +31,9 @@
 
 /**
  * Class EditRootCategoryCommand edits given root category.
+ *
+ * @todo: "root" keyword should be replaced by "home" to avoid confusion between the actual "root" category and "home" category.
+ *         This command is actually handling th Home category edition, as the actual "root" category doesn't exist from UX perspective
```
Comment (jolelievre) [inline src/Core/Domain/Category/Command/EditRootCategoryCommand.php]:
```suggestion
 *         This command is actually handling the Home category edition, as the actual "root" category doesn't exist from UX perspective
```

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: I think 0 is a valid value when you associate no tax to the product, and `TaxRulesGroupId` accepts the 0 value

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php [hunk 10]
```diff
@@ -245,101 +228,214 @@ public function categoryDoesNotExist(string $categoryReference)
...
+     * @param EditCategoryCommand|EditRootCategoryCommand $command
+     * @param TableNode $tableNode
      *
-     * @param string $categoryReference
-     * @param TableNode $table
+     * @see EditCategoryCommand
+     * @see EditRootCategoryCommand
+     *
+     * Technically both commands should be filled separately,
+     * but it happens to match almost all properties except $parentId, so we can reuse them here.
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php:375]: 👍

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php [hunk 11]
```diff
@@ -245,101 +228,214 @@ public function categoryDoesNotExist(string $categoryReference)
...
+     *
+     * Probably these commands are implemented wrong and instead should have extended each other if they supposed to share the properties and the logic,
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]: Actually, I believe extending commands is a bad practice, each command is supposed to be independent and should define its own clear contract. Because even if it's done with the idea of avoiding duplication in mind, from a technical point of view it creates a coupling between the two commands, modifying the common parent will impact both commands even if the new field is not related to both. You can always stop sharing the common parent at that point, but I think that's why it should be avoided.

I guess only in very rare use cases can it be acceptable, for example in the currency we used it because Unofficial currencies have the exact same structure as official currencies, they just obey different domain rules which are handled by the Handler itself depending on which command you use. But even then I struggled a long time to decide if I should use a common parent or not, and I still have doubts about it 😅

File: tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature [hunk 12]
```diff
@@ -1,165 +1,752 @@
...
+      | active              | false            |
+      | parent category     | home             |
+      | link rewrite[en-US] | mobile-phones-en |
+      | link rewrite[fr-FR] | mobile-phones-fr |
+    When I delete category "category3" choosing mode "associate_and_disable"
+    Then category "category3" does not exist
+    When I delete category "category4" choosing mode "associate_only"
+    Then category "category4" does not exist
+    When I delete category "category5" choosing mode "remove_associated"
+    Then category "category5" does not exist
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature:158]: It doesn't really test the effect of each mode, we'd need to associate products to check this But I think it should be handled in a dedicated PR and a dedicated feature file At least it validates that each mode is usable 😅

File: tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature [hunk 13]
```diff
@@ -1,165 +1,752 @@
...
+
+  Scenario: Delete a category which is the last category of that product
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature:160]: Oh, actually my bad you already tested it!! Very nice!

File: tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature [hunk 14]
```diff
@@ -1,165 +1,752 @@
...
+      | id reference     | name             | is default |
+      | home-accessories | Home Accessories | true       |
+    Then product "product_b2" should be assigned to following categories:
+      | id reference     | name             | is default |
+      | home-accessories | Home Accessories | true       |
+    And product "product_b1" should be enabled
+    And product "product_b2" should be enabled
+
+  Scenario: Bulk delete categories which are the last remaining associated categories for some product
+  by using associate_and_disable deletion mode
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature]:
```suggestion
  Scenario: Bulk delete categories which are the last remaining associated categories for some product by using associate_and_disable deletion mode
```
Scenario details should remain on the same line no? Or is it accepted by Gerkhin syntax?

File: tests/Integration/Behaviour/behat.yml [hunk 15]
```diff
@@ -63,6 +63,14 @@ default:
...
                 - Tests\Integration\Behaviour\Features\Context\CategoryFeatureContext
                 - Tests\Integration\Behaviour\Features\Context\Domain\CategoryFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Domain\Product\UpdateCategoriesFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Domain\Product\AddProductFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Domain\Product\UpdateStatusFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Domain\Product\CommonProductFeatureContext
+                - Tests\Integration\Behaviour\Features\Transform\StringToBoolTransformContext
+                - Tests\Integration\Behaviour\Features\Context\ShopFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\LanguageFeatureContext
+                - Tests\Integration\Behaviour\Features\Context\Domain\CustomerFeatureContext
```
Comment (jolelievre) [inline tests/Integration/Behaviour/behat.yml]: Can you please order the contexts alphabetically, I struggled re-ordering them all a while back 😅 Maybe in the meantime some have been added not in the right place, but if we can keep them correctly ordered it reduces the chances of conflicts

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

Amazing job!! And great work with the behat tests

File: tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature (hunk 14 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature]: it works, so I thought why not. Easier to read :smile:

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 6 above)
Comment (lartist) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: Ok if it's considered as a valid value

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php (hunk 5 above)
Comment (zuk3975) [reply src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php:191]: the scenario you are talking about has a related deletion mode to it :arrow_up: (you can choose to remove the products).
But the one Im mentioning is this https://github.com/PrestaShop/PrestaShop/pull/30231/files#diff-6522f54552ae3b1e5495e7835ccf33da060205734c9dd07368b8af46f5a7e5f1R102
previously it would delete product if parent category id 0 is provided. So idk, just seems to be abit random :smile:

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 6 above)
Comment (FabienPapet) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: We should never handle `0` values (and sorry about that this is the case here) .If the TaxGroupRuleId does not exists, I think it's better to throw a `RuntimeException` if the `TaxGroupRuleId` should always exist, otherwise a proper exception would be better appropriate.

Comment (zuk3975) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: Just to make it clear, Im not fixing it in this PR, because its absolutely out of scope, so as it was your change, @FabienPapet (at least I see that by history), I trust it to you to fix in a dedicated PR :stuck_out_tongue: 
And yes, if TaxRulesGroupId is optional value it should insipre from similar cases like https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Domain/Manufacturer/ValueObject/NoManufacturerId.php.
Else if its not optional, then this method should throw TaxRulesGroupIdNotFoundException.

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php (hunk 11 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]: Yes it depends on what we expect. If we expect these commands going separate ways, then yes, we shouldn't have a shared contract for them.
> for example in the currency we used it because Unofficial currencies have the exact same structure as official
> currencies, they just obey different domain rules which are handled by the Handler

It seems very similar concept here to me and quite a simple example of inheritence. The "home" category is still a category, the only difference, that we don't want to manipulate its parentId, everything else seems to be the same. (its 90% chance they will both evolve together) :man_shrugging: Ofc its debatable and it is already done so nvm... Even the basic inheritence examples seems ok at first, but something new comes up and you know you failed - like `car extends vehicle - it has wheels, steering wheel, seats. Everything works and seems nice when one day there comes some tesla tank which has tracks instead of wheels and is self-driven by AI so there is neither seats nor the steering wheel `:sweat_smile: ).
Anyway, I'll adjust the comment abit, but I will leave the "reusable" method here as I think for tests its ok. And its easy to separate it to dedicated methods for each command later if we see that it gets too messy.

File: tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature (hunk 12 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature:158]: yep this was just a basic test to see that all modes deletes the category. And as you've already seen bellow I implemented dedicated cases to test every method when the category is the last for the product.

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 7 above)
Comment (zuk3975) [reply src/Adapter/Product/Repository/ProductRepository.php]: ok waiting 8.0.x -> develop merge to use the trait.

**Comment by zuk3975** [comment]
reopening after sanity issues have been fixed

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 7 above)
Comment (zuk3975) [reply src/Adapter/Product/Repository/ProductRepository.php]: ok so after it was finally merged I checked more in depth and the trait method requires ShopConstraint it means I would need to refacto both categoryDelete handlers to add a shop constraint (which would also introduce BC break), but that also I would basically have to refacto whole categories deletion to support multistore :sweat: That would require additional multistore tests, handle retrieving/deleting correct shops products for categories, deleting categories based on shop constraint etc... Altough it must be done at some point, but Im afraid it will be too much for a single PR :sweat: 
Maybe to avoid polluting the CategoryRepository, I should just leave those plain sql's in the abstract handler as it was?

Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductRepository.php]: Ok. let's reduc the refacto on this PR then. And maybe you shouldn't even add this in the repository Let's save this refacto for later and keep the handler's modifications to the minimum

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: I agree in other use cases we used a common interface a `NoValueObject` dedicated classes for the 0 value But it's not in this PR's scope and can be improved in another PR (it will also create BC since `TaxRulesGroupId` will no longer accept the 0 value)

The `getProductTaxRulesGroupId` should not trigger an `TaxRulesGroupIdNotFoundException` though since the 0 value is acceptable, it should return a VO matching the interface that could be either `TaxRulesGroupId` or `NoTaxRulesGroupId`

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 6 above)
Comment (lartist) [reply src/Adapter/Product/Repository/ProductRepository.php:426]: I'm ok but plz let's think about fixing it soon.

**Review by lartist — APPROVED** [review]

**Comment by AureRita** [comment]
Hi @zuk3975 

Thank you for your PR, I tested it and it  works, as we can see here :

[Untitled_ Dec 2, 2022 4_44 PM.webm](https://user-images.githubusercontent.com/100699445/205334035-f1abced1-7a4a-4841-8253-59b712875e22.webm)

[Untitled_ Dec 2, 2022 5_05 PM.webm](https://user-images.githubusercontent.com/100699445/205335457-faa2cc7d-32e4-47d9-9fc4-f7adec59f303.webm)


Thanks

**Review by AureRita — APPROVED** [review]

**Comment by marsaldev** [comment]
Thanks @zuk3975 ! 💪
